data/ipfs.go

```diff
@@ -257,7 +262,76 @@ func (i *IPFSHandle) Retrieve(ctx context.Context, path string, maxSize int64) (
 	}
 	r, ok := node.(files.File)
 	if !ok {
-		return nil, errors.New("received incorrect type from Unixfs().Get()")
+		return nil, fmt.Errorf("received incorrect type from Unixfs().Get()")
 	}
 	return io.ReadAll(r)
 }
+
+// PublishIPNSpath creates or updates an IPNS record with the content of a
+// filesystem path (a single file or a directory).
+//
+// The IPNS record is published under the scope of the private key identified
+// by the keyalias parameter. New keys can be created using method AddKeyToKeystore
+// and function NewIPFSkey() both available on this package.
+//
+// The execution of this method might take a while (some minutes),
+// so the caller must handle properly the logic by using goroutines, channels or other
+// mechanisms in order to not block the whole program execution.
+func (i *IPFSHandle) PublishIPNSpath(ctx context.Context, path string,
+	keyalias string) (coreiface.IpnsEntry, error) {
+	root, err := i.AddAndPin(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	c, err := cid.Parse(root)
+	if err != nil {
+		return nil, err
+	}
+	if keyalias == "" {
+		ck, err := i.CoreAPI.Key().Self(ctx)
+		if err != nil {
+			return nil, err
+		}
+		keyalias = ck.Name()
+	}
+	return i.CoreAPI.Name().Publish(
+		ctx,
+		ipath.IpfsPath(c),
+		options.Name.TTL(time.Duration(time.Minute*10)),
+		options.Name.Key(keyalias),
+	)
+}
+
+// AddKeyToKeystore adds a marshaled IPFS private key to the IPFS keystore.
+// The key is identified by a unique alias name which can be used for referncing
+// that key when using some other IPFS methods.
+// Compatible Keys can be generated with NewIPFSkey() function.
+func (i *IPFSHandle) AddKeyToKeystore(keyalias string, privkey []byte) error {
+	pk, err := ipfscrypto.UnmarshalPrivateKey(privkey)
+	if err != nil {
+		return err
+	}
+	_ = i.Node.Repo.Keystore().Delete(keyalias)
+	if err := i.Node.Repo.Keystore().Put(keyalias, pk); err != nil {
+		if err != keystore.ErrKeyExists {
+			return err
+		}
+	}
+	return nil
+}
+
+// NewIPFSkey generates a new IPFS private key (ECDSA/256bit) and returns its
+// marshaled bytes representation.
+func NewIPFSkey() []byte {
+	// These functions must not return error since all input parameters
+	// are predefined, so we panic if an error returned.
+	pk, _, err := ipfscrypto.GenerateKeyPair(ipfscrypto.ECDSA, 256)
```

Comment:
I had to do a double take to think if pk was "private key" or "public key"; I'd maybe use `privKey` to be explicit. And if you have a marshaled version, then `encPrivKey` or something like that.